qubes/mgmt.py

```diff
@@ -186,11 +234,239 @@ def vm_property_help(self, untrusted_payload):
         return qubes.utils.format_doc(doc)
 
     @asyncio.coroutine
-    def vm_property_reset(self, untrusted_payload):
+    @no_payload
+    def vm_property_reset(self):
         assert self.arg in self.dest.property_list()
-        assert not untrusted_payload
-        del untrusted_payload
 
         self.fire_event_for_permission()
 
         delattr(self.dest, self.arg)
+        self.app.save()
+
+    @asyncio.coroutine
+    @no_payload
+    def vm_volume_list(self):
+        assert not self.arg
+
+        volume_names = self.fire_event_for_filter(self.dest.volumes.keys())
+        return ''.join('{}\n'.format(name) for name in volume_names)
+
+    @asyncio.coroutine
+    @no_payload
+    def vm_volume_info(self):
+        assert self.arg in self.dest.volumes.keys()
+
+        self.fire_event_for_permission()
+
+        volume = self.dest.volumes[self.arg]
+        # properties defined in API
+        volume_properties = [
+            'pool', 'vid', 'size', 'usage', 'rw', 'internal', 'source',
+            'save_on_stop', 'snap_on_start']
+        return ''.join('{}={}\n'.format(key, getattr(volume, key)) for key in
+            volume_properties)
+
+    @asyncio.coroutine
+    @no_payload
+    def vm_volume_listsnapshots(self):
+        assert self.arg in self.dest.volumes.keys()
+
+        self.fire_event_for_permission()
+
+        volume = self.dest.volumes[self.arg]
+        return ''.join('{}\n'.format(revision) for revision in volume.revisions)
+
+    @asyncio.coroutine
+    def vm_volume_revert(self, untrusted_payload):
+        assert self.arg in self.dest.volumes.keys()
+        untrusted_revision = untrusted_payload.decode('ascii').strip()
+        del untrusted_payload
+
+        volume = self.dest.volumes[self.arg]
+        snapshots = volume.revisions
+        assert untrusted_revision in snapshots
+        revision = untrusted_revision
+
+        self.fire_event_for_permission(revision=revision)
+
+        self.dest.storage.get_pool(volume).revert(revision)
+        self.app.save()
+
+    @asyncio.coroutine
+    def vm_volume_resize(self, untrusted_payload):
+        assert self.arg in self.dest.volumes.keys()
+        untrusted_size = untrusted_payload.decode('ascii').strip()
+        del untrusted_payload
+        assert untrusted_size.isdigit()  # only digits, forbid '-' too
+        assert len(untrusted_size) <= 20  # limit to about 2^64
```

Comment:
This is dangerously close to current filesystem limits. For example ext4 has a maximum volume size limit of 1EiB = 2^60 B. ``len(str(2**60)) == 19``.